Fix crash in muon analysis fitting interface

jclarkeSTFC opened the pull request:
The bug described in issue #33505 is an exception thrown in the destructor of either CreateWorkspace or EvaluateFunction. EvaluateFunction was being called directly in the muon analysis fitting GUI without going through the AlgorithmManager, which I think is the source of the problem. So I used the AlgorithmManager instead.

**Description of work**
- Used AlgorithmManager for calling EvaluateFunction
- Moved creation of workspace into smaller scope

**Purpose of work**
<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
-->
Fixes #33505 

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**Summary of work**
<!-- Please provide a short, high level description of the work that was done.
-->

**To test:**
Original bug in #33505, instructions to reproduce (they're from a comment further down the page):
1. Open Muon Analysis GUI
2. Load EMU00020884.nxs from the Training Course Data set
3. Click on the fitting tab
4. Add a Fitting function (e.g, Dynamic Kubo Toyabe)
5. Tick Plot guess
6. Untick plot guess
7. Add another fitting function
8. Click plot guess

There are other combinations of adding/removing fitting functions and plotting guesses that resulted in the same thing, so when testing probably best to try various combos.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Use algorithm manager for EvaluateFunction.

The bug described in issue #33505 is an exception
thrown in the destructor of either CreateWorkspace or
EvaluateFunction. EvaluateFunction was being called
directly in the muon analysis GUI without going
through the AlgorithmManager, which I think is the
source of the problem. So I used the AlgorithmManager
instead.

jclarkeSTFC committed:
Added release notes.

AnthonyLim23 commented:
The failure is because the mocks/patches need updating e.g. https://github.com/mantidproject/mantid/blob/main/qt/python/mantidqtinterfaces/test/Muon/fitting_widgets/basic_fitting/basic_fitting_model_test.py#L510

jclarkeSTFC committed:
Update mocks in two tests.

EvaluateFunction is no longer called when evaluating the plot guess, but evaluate_function instead.

jhaigh0 reviewed (approved):
Instructions no longer produce an error and code looks good 👍
